katcp/testutils.py

```diff
@@ -1809,6 +1810,39 @@ def assert_wait_call_count(self, count, timeout=1.):
         if self.call_count >= count and len(self.call_args_list) < count:
             raise RuntimeError("call_args_list not updated within timeout.")
 
+class AsyncWaitingMock(mock.Mock):
+
+    def __init__(self, *args, **kwargs):
+        super(AsyncWaitingMock, self).__init__(*args, **kwargs)
+        self._call_event = tornado.locks.Event()
+
+    def _mock_call(self, *args, **kwargs):
+        ret = super(AsyncWaitingMock, self)._mock_call(*args, **kwargs)
+        self._call_event.set()
+        return ret
+
+    @tornado.gen.coroutine
+    def assert_wait_call_count(self, count, timeout=1.):
+        """Wait for mock to be called >= *count* times within *timeout* seconds.
+
+        Raises AssertionError if the call count is not reached.
+
+        """
+        t0 = time.time()
+        if self.call_count >= count and len(self.call_args_list) >= count:
+            raise tornado.gen.Return(True)
+        to_wait = timeout
+
+        while to_wait >= 0 and self.call_count < count:
+            try:
+                yield self._call_event.wait(timeout)
+                self._call_event.clear()
+            except gen.TimeoutError:
```

Comment:
I'm not sure what the attribute errors are about. I don't see them when I run locally. It only happens after the actual nosetests are done, which makes me think that there may be some daemon (shudder) threads running at the end of the tests getting tangled up when the python process tears itself down. See https://joeshaw.org/python-daemon-threads-considered-harmful/